sql: generate spans for filters with OR

RaduBerinde opened the pull request:
Note -- the first commit is a cosmetic test change, I recommend reviewing commit 2 separately.

This change adds basic support for top-level ORs in filters. Previously only a
few simplifications were supported, e.g. `a = 1 OR a = 3` gets converted to `a
in (1,2)`.

We don't support using multiple indexes in parallel the same time; we take into
account all disjunctions with respect to the same index. We generate the
indexConstraints for each disjunction; we then generate spans for each
disjunction separately, and finally merge the spans.

This works for a large class of expressions (with the proper index), like
- `a < 5 OR a > 10`
- `(a >= 1 AND a <= 10) OR (a >= 20 AND a <= 30)`
- `(a = 1 AND b > 10) OR (a = 5 AND b < 10)`

Expressions like `a > 1 OR b > 1` will still not generate any constraints (we
would have to support joining from multiple indexes).

Fixes #5189.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5521)

<!-- Reviewable:end -->

tamird commented:
> a = 1 OR a = 3` gets converted to `a in (1,2)`

doesn't sound right

> We don't support using multiple indexes in parallel the same time

**at** the same time

---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[sql/index_selection.go, line 150 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDaRHEJSDZNpgqgQgC-:-KDaRHEKqxe7aO8d50Oy:112752660) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L150)):*
double space here

---

*[sql/index_selection.go, line 271 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDaRXfo2vDo07E4csdB:-KDaRXfp4J8pHbA8DHXo:-41395432) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L271)):*
s/Constraint,/Constraints,/

---

*[sql/index_selection.go, line 276 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDaRTdzNNXz8QrehokR:-KDaRTdzNNXz8QrehokS:2027391668) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L276)):*
receiver `oic`?

---

*[sql/index_selection.go, line 330 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDaRi_c1w5dDbfZV7-2:-KDaRi_c1w5dDbfZV7-3:358031074) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L330)):*
kinda prefer the if-else

---

*[sql/index_selection.go, line 334 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDcsGky5ZKPdCYXQ0IR:-KDcsGky5ZKPdCYXQ0IS:-1535653988) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L334)):*
> We thus take the minimum number of restricted columns among them.

Note that other surrounding comments tie back to how the calculated cost is affected - reading this new comment I can't tell what "take" means. Perhaps "We thus calculate the cost based on the disjunction which restricts the smallest number of columns" would read better.

---

*[sql/index_selection.go, line 345 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDcse8Xjjl1AAslRuLE:-KDcse8Xjjl1AAslRuLF:157745540) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L345)):*
nit: yoda condition. prefer `numCols < minNumCols`

---

*[sql/index_selection.go, line 407 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDct9V4P-sCNQ22_YN5:-KDct9V4P-sCNQ22_YN6:-1461714064) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L407)):*
s/coon/con/ let's keep it civil

---

*[sql/index_selection.go, line 417 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDctv7SysWhEMxbNlNd:-KDctv7SysWhEMxbNlNe:19559957) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L417)):*
can this actually happen? I think this comment needs to explain how something like `OR TRUE` winds up as `nil` here

---

*[sql/index_selection.go, line 1049 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDcvtj4_ZyV5ngIzIl6:-KDcvtj5TrAoQwEmRgwO:-2146834162) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L1049)):*
note to self: didn't review from here down

---

*[sql/index_selection.go, line 1287 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDcvNJVXlqQbyQV6FsZ:-KDcvNJVXlqQbyQV6Fs_:-379529800) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L1287)):*
rather than adding this logic, wouldn't you want to simplify the expression first? seems like the cases where exactPrefix isn't zero are those which can be simplified to the point of the disjunctions in them being irrelevant. e.g.

`(a, b) = (1, 2) OR (a, b, c) = (1, 3, 0)` => `a = 1 AND (b = 2 OR (b, c) = (1, 3, 0))`

---

*[sql/index_selection.go, line 1333 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDcv4csNkN_k65HZRjE:-KDcv4csNkN_k65HZRjF:-1124883004) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L1333)):*
move this down to avoid allocating

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-:-KDaQ16vllbQdlg9SZrf:1006103255)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR! Updated.

---

Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions.

---

*[sql/index_selection.go, line 407 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDct9V4P-sCNQ22_YN5:-KDcxuAy-5p_Yf2CcxUL:1807749520) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L407)):*
Hahaha

---

*[sql/index_selection.go, line 417 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDctv7SysWhEMxbNlNd:-KDcxyEHq-zwceSkaGoA:-197744837) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L417)):*
It's actually very common - remember this is all in the context of a certain index. If the condition is `b > 1` and the index is on `a`, we get no constraints.

---

*[sql/index_selection.go, line 1287 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDcvNJVXlqQbyQV6FsZ:-KDcy9edG0oEFq4zNvBQ:-148500310) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L1287)):*
AFAIK we don't have a reliable way to simplify the expression in that way. In any case, I think the results here have to match exactly how we are handling the expression when we make the constraints/spans. And we can't rewrite the expression like that from the get-go because we don't have any way of dealing with `OR`s under `AND`s (other than maybe rewriting the expression to bubble the ORs up which brings us back to the same issue).

Obviously we have a lot of stuff to improve wrt expressions - e.g. it looks like currently (and unrelated to this change) we don't even figure out that `a` is an exact prefix when we have `(a, b) IN ((1, 2), (1, 3))` (I added a TODO about that). On the bright side, if a user hits a problem related to the form of the expression, we can advise an alternate way of passing the expression in SQL.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-:-KDd2-TNJodf3EkqX1U4:-1512379193)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3, 2 of 2 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/index_selection.go, line 417 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDctv7SysWhEMxbNlNd:-KDd4IJIScl8hL3NrT-Q:-303390682) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L417)):*
Ah, it's an empty slice, not `nil`. OK, SGTM.

---

*[sql/index_selection.go, line 1287 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521#-KDcvNJVXlqQbyQV6FsZ:-KDd4xQKjLW9qjD3x6vs:-643657397) ([raw file](https://github.com/cockroachdb/cockroach/blob/53dba3fd383557aa2ef150c1615adf3448dfe714/sql/index_selection.go#L1287)):*
Yeah, I'm not surprised the facility for doing that simplification doesn't exist - I was thinking out loud that perhaps going down the road of implementing it was preferable to this "workaround". In any case, I don't want to block you on this point, just wanted to put those thoughts on your radar.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5521)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: make index selection testcases easier to work with

Instead of defining index columns using string or struct arrays, we use a simple
string encoding, e.g. `a,b-,c` means a (ascending), b (descending), c
(ascending).

RaduBerinde committed:
sql: generate spans for filters with OR

This change adds basic support for top-level ORs in filters. Previously only a
few simplifications were supported, e.g. `a = 1 OR a = 3` gets converted to `a
IN (1,3)`.

We don't support using multiple indexes in parallel at the same time; we take
into account all disjunctions with respect to the same index. We generate the
indexConstraints for each disjunction; we then generate spans for each
disjunction separately, and finally merge the spans.

This works for a large class of expressions (with the proper index), like
 - `a < 5 OR a > 10`
 - `(a >= 1 AND a <= 10) OR (a >= 20 AND a <= 30)`
 - `(a = 1 AND b > 10) OR (a = 5 AND b < 10)`

Expressions like `a > 1 OR b > 1` will still not generate any constraints (we
would have to support joining from multiple indexes).

Fixes #5189.
